README
This is Perl module POEIKC.

INSTALLATION

POEIKC installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan POEIKC

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

POEIKC documentation is available as in POD. So you can do:

    % perldoc POEIKC

to read the documentation online with your favorite pager.

SEE ALSO

    % poeikcd --help
    % poikc   --help

Yuji Suzuki


Changes
Revision history for Perl extension POEIKC

0.03_01   Thu Oct  9 2008
        - added an option of startup. default is "spawn". 
          eg) poeikcd -M MyModule --startup start
                 `MyModule->spawn'
              poeikcd -M MyModule --startup=new start
                 `MyModule->new'

0.03      Tue Sep  2 2008
          The change from 0.02
        - added "POEIKC::Daemon::P2P".  Peer-to-Peer is made in this.
        - revised a logic of shutdown.
        - added the option which changed $0($PROGRAM_NAME).
        - able to do it in "delay" in "loop".

0.02_07   Thu Aug 29 2008
        - added the option which changed $0($PROGRAM_NAME)
        - able to do it in "delay" in "loop".
          eg)
             "poikc -U loop  MyModule::my_method args .."
                package MyModule;
                our $delay = 3;

0.02_06   Thu Aug 28 2008
        - revised a test script.
        - revised a logic of shutdown.
          did shutdown before returning a response to a client, but was able 
          to give a client the time to spare that gave back a response.

0.02_05   Thu Aug 27 2008
        - revised a test script.
        - corrected a document.

0.02_04   Thu Aug 25 2008
        - revised a test script.

0.02_03   Thu Aug 21 2008
        - added a test of "POEIKC::Daemon::P2P"

0.02_02   Mon Aug 18 2008
        - "POEIKC::Daemon::AndClient" changed a name in "POEIKC::Daemon::P2P".
           As for poeikcd, the network constitution of the Peer-to-Peer model 
           was enabled.

0.02_01   Sun Aug 17 2008
        - Added "POEIKC::Daemon::AndClient"(fictitious name) to be able to 
          easily communicate between poeikcd.

0.02     Wed Aug 13 2008
        - added the option which could change a server name. 
          By this, I can communicate between two server & clients!

0.01     Thu Feb 27 2008
        - The official version

0.00_07  Thu Feb 22 2008
        - poikc : -T option. history resize.
        - The revision of the 'Makefile.PL'. recommends 'Term::ReadLine::Gnu';

0.00_06  Thu Feb 21 2008
        - poikc : -T option. decided to use 'Term::ReadLine::Gnu'.

0.00_05  Thu Feb 21 2008
        - poeikcd : -I option. can put PATH at the time of the start in @INC.
            eg) -I ./lib 
        - poikc : -T option. Added a history command.
                  history file path : $ENV{HOME}.'/.poikc_history'

0.00_04  Thu Feb 20 2008
        - poikc : -T option. got possible to almost use the terminal. 
                             supported an argument including the blank.
        - poikc : -T option. Added a use command.
            eg) 
            $ poikc -T
            localhost:47225 POEIKCd> use LWP::Simple
            localhost:47225 POEIKCd> ::get http://search.cpan.org/~suzuki/

0.00_03  Thu Feb 19 2008
        - The revision of the 'Makefile.PL'.
        - Directory rearranging in the inc directory.

0.00_02  Thu Feb 16 2008
        - The revision of the test cord.

0.00_01  Thu Feb 15 2008
        - poeikcd : supported a parameter of IKC server.
        - poikc : Added -T option. took in a terminal experimentally. 
          There is malfunction.

0.00_00  Thu Feb 14 00:57:56 2008
        - replaced from POEIKCdaemon to POEIKC.


eg/client_lite.pl
#!/usr/local/bin/perl

###	perl -I t -I lib -MPOEIKC::Daemon -e POEIKC::Daemon::daemon

use strict;
use warnings;
$| = 1;
use POE::Component::IKC::ClientLite;
use Data::Dumper;
use Sys::Hostname;
use Getopt::Long;
use Pod::Usage;

my $options = {};

BEGIN {
	$Getopt::Long::ignorecase=0;
	$options = {};
	GetOptions ($options, qw/
		HOST=s 
		port=i 
		alias=s
		debug
		help
		/);
}
	my $menu = shift;

	$options->{help} and pod2usage(1);

	$options->{alias} ||= 'POEIKCd';
	$options->{port} ||= 47225;
	$options->{state_name} ||= 'method_respond';
	$options->{HOST} ||= '127.0.0.1';
	$options->{state_name} = 
		$options->{state_name} =~ /^m/i ? 'method_respond' : 
		$options->{state_name} =~ /^f/i ? 'function_respond' : pod2usage(1);

	print scalar localtime,"\n";
	printf "[poeikcd ..  %s / PORT:%s]\n", $options->{HOST}, $options->{port};

	my ($name) = $0 =~ /(\w+)\.\w+/;
	$name .= $$;
	my $ikc = create_ikc_client(
			ip => $options->{host},
			port => $options->{port},
			name => $name,
	);

	$ikc or do{
		printf "%s\n\n",$POE::Component::IKC::ClientLite::error; 
		pod2usage(1);
	};

	my $ret;
	my $nom=0;
	my %exe = map {$nom++=>$_}
	(
		['Foo::Class'=>'FooMethod'=>'@args'],
		['POEIKC::Daemon::Utility' => 'reload', 'POEIKC::Daemon::Utility'],
		['POEIKC::Daemon::Utility' => 'get_H_ENV'],
		['POEIKC::Daemon::Utility' => 'get_A_INC'],
		['POEIKC::Daemon::Utility' => 'unshift_INC', './t'],
		['POEIKC::Daemon::Utility' => 'unshift_INC', '~/lib'],
		['POEIKC::Daemon::Utility' => 'delete_INC', './t'],
		['POEIKC::Daemon::Utility' => 'reset_INC'],
		['POEIKC::Daemon::Utility' => 'get_H_INC'],
		['POEIKC::Daemon::Utility' => 'get_pid'],
		['POEIKC::Daemon::Utility' => 'get_object_something', 'ikc_self_port'],
		['POEIKC::Daemon::Utility' => 'get_object_something', 'session_alias'],
		['POEIKC::Daemon::Utility' => 'get_stay'],
		['POEIKC::Daemon::Utility' => 'get_VERSION'],
		['POEIKC::Daemon::Utility' => 'stop'],
		['Cwd' => 'getcwd'],
		['IKC_d_Localtime' => 'timelocal'],
		['POEIKC::Daemon::Utility' => 'reload', 'IKC_d_Localtime'],
		['POEIKC::Daemon::Utility' => 'reload', 'IKC_d_Localtime' => 'timelocal'],
		['POEIKC::Daemon::Utility' => 'stay', 'IKC_d_Localtime' ],
		['POEIKC::Daemon::Utility' => 'get_Class_Inspector', 'POEIKC::Daemon::Utility'],
		['POEIKC::Daemon::Utility' => 'get_Class_Inspector', 'POEIKC::Daemon::Utility','methods'],
		['Cwd' => 'getcwd'],
		['LWP::Simple' => 'get', 'http://search.cpan.org/~suzuki/'],
	);


	printf "%2d => %s\n", $_, join("\t"=> @{$exe{$_}}) for sort {$a<=>$b} keys %exe;

	print '*' x 20, "\n";

	$menu or printf("\n    perl eg/client_lite.pl [1 .. %d]\n\n", scalar(keys %exe)-1), exit;

	printf "[%d]\t%s\n", $menu, join "\t"=>@{$exe{ $menu }};

	my $session_alias = $options->{alias} || 'POEIKCd';
	
	my $event = $menu <= 15 ? 'method_respond' : 'function_respond';
	print $event,"\n";

	$ret = $ikc->post_respond($session_alias.'/'.$event => $exe{ $menu });

	$ikc->error and die($ikc->error);
	if (my $r = ref $ret) {
		if ( $r eq 'HASH'){
			my %ret = %{$ret};
			for(sort keys %ret){printf "%-35s= %s", $_, Dumper $ret{$_}}
			print "\n";
		}else{
			print(Dumper($ret));
		}
	}else{
		print(Dumper($ret));
	}

__END__

=head1 SYNOPSIS

  Options:

	-H  --HOST=s        : default 127.0.0.1 

    -p  --port=#        : Port number to use for connection.
                          default 47224 

    -a  --alias=s       : session alias
                          default POEIKCd 

    -h  --help

=cut
